src/git/repo.rs

```diff
@@ -965,6 +1022,144 @@ Either create it, or update the main branch setting by running:
+        let changed_paths = {
+            let mut result: HashSet<&Path> =
+                parent_commit_pathbufs.iter().map(PathBuf::borrow).collect();
+            let paths = match opts {
+                AmendFastOptions::FromIndex { paths } => paths,
+                AmendFastOptions::FromWorkingTree { paths } => paths,
+            };
+            result.extend(paths.iter().map(|path| path.as_path()));
+            result.iter().cloned().collect_vec()
+        };
+
+        let dehydrated_parent =
+            self.dehydrate_commit(parent_commit, changed_paths.as_slice(), true)?;
+        let dehydrated_parent_index = Index::from_tree(&dehydrated_parent.get_tree()?)?;
+
+        let repo_path = self
+            .get_working_copy_path()
+            .expect("repo working copy path");
+        let amended_oids: HashMap<PathBuf, Option<NonZeroOid>> = match opts {
+            AmendFastOptions::FromWorkingTree { paths } => paths
+                .iter()
+                .map(|path| {
+                    let file_path = &repo_path.join(path);
+                    if file_path.exists() {
+                        // Create a new blob OID based on the current on-disk
+                        // contents of the file in the working copy.
+                        self.inner
+                            .blob_path(file_path)
+                            .map(|oid| (path.clone(), Some(make_non_zero_oid(oid))))
+                    } else {
+                        Ok((path.clone(), None))
+                    }
+                })
+                .collect::<Result<HashMap<_, _>, _>>()?,
+            AmendFastOptions::FromIndex { paths } => {
+                let index = self.get_index()?;
+                paths
+                    .iter()
+                    .filter_map(|path| match index.get_entry(path) {
+                        Some(IndexEntry {
+                            oid: MaybeZeroOid::Zero,
+                            ..
+                        }) => {
+                            warn!("index entry for {:?} was zero", path);
+                            None
+                        }
+                        Some(IndexEntry {
+                            oid: MaybeZeroOid::NonZero(oid),
+                            ..
+                        }) => Some((path.clone(), Some(oid))),
+                        None => Some((path.clone(), None)),
+                    })
+                    .collect::<HashMap<_, _>>()
+            }
+        };
+
+        let amended_index_entries: HashMap<PathBuf, Option<(NonZeroOid, i32)>> = changed_paths
+            .into_iter()
+            .map(|changed_path| {
+                let value = if let Some(amended_oid) = amended_oids.get(changed_path) {
+                    match amended_oid {
+                        Some(added_blob_oid) => {
+                            let file_path = repo_path.join(changed_path);
+                            let file_mode = if file_path.is_executable() {
```

Comment:
Rather than read from disk again in a potentially racy way to determine if the file mode is executable, it would probably be better to use the `--porcelain=v2` format to `git status` (assuming that it's supported by Git v2.24, our lowest-supported version), which includes this information.